doc/sphinx/proofs/writing-proofs/proof-mode.rst

```diff
@@ -1,66 +1,165 @@
+Coq enters proof mode when you begin a proof,
+such as with the :cmd:`Theorem` command.  It exits proof mode when
+you complete a proof, such as with the :cmd:`Qed` command.  Tactics,
+which are available only in proof mode, incrementally transform incomplete
+proofs to eventually generate a complete proof.
 
-Coq user interfaces usually have a way of marking whether the user has
-switched to proof editing mode. For instance, in coqtop the prompt ``Coq <``   is changed into
-:n:`@ident <`   where :token:`ident` is the declared name of the theorem currently edited.
+When you run Coq interactively, such as through CoqIDE, Proof General or
+coqtop, Coq shows the current proof state--the incomplete proof--as you
+enter tactics.  This information isn't shown when you run Coq in batch
+mode with `coqc`.
 
-At each stage of a proof development, one has a list of goals to
-prove. Initially, the list consists only in the theorem itself. After
-having applied some tactics, the list of goals contains the subgoals
-generated by the tactics.
+Proof State
+-----------
 
-To each subgoal is associated a number of hypotheses called the *local context*
-of the goal. Initially, the local context contains the local variables and
-hypotheses of the current section (see Section :ref:`gallina-assumptions`) and
-the local variables and hypotheses of the theorem statement. It is enriched by
-the use of certain tactics (see e.g. :tacn:`intro`).
+.. todo PR: my example is pretty bad--can you suggest a better one?
+   Also need to position/integrate this better
 
-When a proof is completed, the message ``Proof completed`` is displayed.
-One can then register this proof as a defined constant in the
-environment. Because there exists a correspondence between proofs and
-terms of λ-calculus, known as the *Curry-Howard isomorphism*
-:cite:`How80,Bar81,Gir89,H89`, Coq stores proofs as terms of |Cic|. Those
-terms are called *proof terms*.
+The :gdef:`proof state`, also referred to as the :gdef:`proof context`, consists of one
+or more unproven goals.
+Each goal has a :gdef:`conclusion` (the statement that is to be proven)
+and may have zero or more named facts that can be used in proving the
+conclusion (in addition to definitions, proven theorems, etc.).
 
+The term ":gdef:`goal`" may refer to an entire goal or to the conclusion
+of a goal, depending on the context.
 
-.. exn:: No focused proof.
+The facts appear above a line and the conclusion appears below that line.
+They are both types.  Facts can be :gdef:`hypotheses <hypothesis>`
+(which are propositions), variables or local definitions.
+The name of the fact appears first.  The type is shown after the colon.  Local
+definitions are shown in the form `n := 0 : nat`.
 
-   Coq raises this error message when one attempts to use a proof editing command
-   out of the proof editing mode.
+The proof state when you begin proving a theorem shows
+the statement of the theorem below the line and no hypotheses above the
+line:
```

Comment:
Also known as an empty proof context.

Reply:
Sure.  But probably not worth introducing the term unless we use it more than once.